onchain/rollups/test/foundry/consensus/quorum/QuorumFactory.t.sol

```diff
@@ -0,0 +1,166 @@
+    }
+
+    function testNewQuorumDeterministic(
+        uint256 _numValidators,
+        bytes32 _salt
+    ) public {
+        vm.assume(_numValidators > 1);
+        vm.assume(_numValidators < 50);
+
+        address[] memory quorumValidators = generateValidators(_numValidators);
+        uint256[] memory shares = generateShares(quorumValidators);
+
+        IHistory history = new History(msg.sender);
+
+        address precalculatedAddress = factory.calculateQuorumAddress(
+            quorumValidators,
+            shares,
+            history,
+            _salt
+        );
+
+        vm.recordLogs();
+
+        Quorum quorum = factory.newQuorum(
+            quorumValidators,
+            shares,
+            history,
+            _salt
+        );
+
+        emit QuorumCreated(quorumValidators, quorum);
+
+        // Precalculated address must match actual address
+        assertEq(precalculatedAddress, address(quorum));
+        decodeFactoryLogs(quorumValidators, quorum);
+    }
+
+    function testAlreadyDeployedNewQuorumDeterministic(
+        uint256 _numValidators,
+        bytes32 _salt
+    ) public {
+        vm.assume(_numValidators > 1);
+        vm.assume(_numValidators < 50);
+
+        address[] memory quorumValidators = generateValidators(_numValidators);
+        uint256[] memory shares = generateShares(quorumValidators);
+
+        IHistory history = new History(msg.sender);
+
+        factory.newQuorum(quorumValidators, shares, history, _salt);
+
+        //Deploy already deployed quorum
+        vm.expectRevert();
+        factory.newQuorum(quorumValidators, shares, history, _salt);
+    }
+
+    // HELPER FUNCTIONS
+    function generateValidators(
+        uint256 _numValidators
+    ) internal pure returns (address[] memory) {
+        address[] memory validators = new address[](_numValidators);
+        for (uint256 i = 0; i < _numValidators; i++) {
+            validators[i] = vm.addr(i + 1);
+        }
+        return validators;
+    }
+
+    function generateShares(
+        address[] memory validators
+    ) internal pure returns (uint256[] memory) {
+        //generate a random number of shares for each validator
+        uint256[] memory shares = new uint256[](validators.length);
+        for (uint256 i; i < shares.length; ++i) {
+            uint256 share = uint256(
+                keccak256(abi.encodePacked(i, validators[i]))
+            ) % 100;
+            shares[i] = (share > 0) ? share : validators.length;
+        }
+        return shares;
+    }
+
+    function decodeFactoryLogs(
+        address[] memory _quorumValidators,
+        Quorum quorum
+    ) internal {
+        Vm.Log[] memory entries = vm.getRecordedLogs();
+
+        for (uint i = 0; i < entries.length; i++) {
+            Vm.Log memory entry = entries[i];
+
+            if (
+                entry.topics[0] == QuorumCreated.selector &&
+                entry.emitter == address(factory)
+            ) {
```

Comment:
We should also check that exactly one event of this type of emitted.
See the code from `CartesiDAppFactory` for reference:

https://github.com/cartesi/rollups-contracts/blob/6f4511976fb3c7a0cff0c4fabbfaa558abec5cb3/onchain/rollups/test/foundry/dapp/CartesiDAppFactory.t.sol#L138-L174